js/store.js

```diff
@@ -131,32 +258,95 @@ const store = {
       throw new Error('setThirdParty requires a valid origin argument');
     }
 
-    const firstParty = this.getWebsite(origin);
-    const thirdParty = this.getWebsite(target);
+    let firstParty = this.getWebsite(origin);
+    let thirdParty = this.getWebsite(target);
+
+/*
+  if not {},
+
+  firstParty is expected to match this format:
+  {
+    thirdPartyHostnames: [
+      "www.thirdpartydomain.com",
+      ...
+    ]
+  }
+  thirdParty is expected to match this format:
+  {
+    firstPartyHostnames: [
+      "www.firstpartydomain.com",
+      ...
+    ],
+    isVisible: false,
+    whiteListedFirstParty: "www.google.com" (optional)
+  }
+  */
 
     if (!('thirdPartyHostnames' in firstParty)) {
       firstParty['thirdPartyHostnames'] = [];
     }
     if (!firstParty['thirdPartyHostnames'].includes(target)) {
-      firstParty['thirdPartyHostnames'].push(target);
-      newThirdParty = true;
+      // third party is not currently linked to first party
+      ({ firstParty, thirdParty, newThirdParty }
+        = this.whiteListCheck(origin, target, firstParty, thirdParty));
     }
-    if (!('firstPartyHostnames' in firstParty)) {
+    if (!('firstPartyHostnames' in thirdParty)) {
       thirdParty['firstPartyHostnames'] = [];
     }
     if (!thirdParty['firstPartyHostnames'].includes(origin)) {
       thirdParty['firstPartyHostnames'].push(origin);
-      newThirdParty = true;
     }
 
     this.setWebsite(origin, firstParty);
     this.setWebsite(target, thirdParty);
 
     if (newThirdParty) {
-      this.updateChild(this.outputWebsite(target, data, origin));
+      this.updateChild(this.outputWebsite(target, data));
     }
   },
 
+  // Checks if third party is on whitelist for first party,
+  // and if the whitelisted third party should be unwhitelisted.
+  whiteListCheck(origin, target, firstParty, thirdParty) {
+    let newThirdParty = false;
+    // first parties are always visible
+    if (!('isVisible' in firstParty)) {
```

Comment:
I don't think we need this key always present in firstParty do we? Especially as this won't be called for first parties when they are new this would be unreliable.